### PR TITLE
Wait for document ready before search for element in #getElement

### LINKED ISSUE #9370: Analytics root: wait for document ready before select element

### PR DESCRIPTION
**Pull request by zhouyx**
fix #9370

### CODE REVIEW

File: extensions/amp-analytics/0.1/analytics-root.js [hunk 1]
```diff
@@ -173,39 +173,45 @@ export class AnalyticsRoot {
...
    *   `'closest'` and `'scope'`.
-   * @return {?Element} Element corresponding to the selector if found.
+   * @return {Promise<?Element>} Element corresponding to the selector if found.
    */
   getElement(context, selector, selectionMethod = null) {
     // Special case selectors. The selection method is irrelavant.
+    // And no need to wait for document ready.
     if (selector == ':root') {
-      return this.getRootElement();
+      return Promise.resolve().then(() => {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/analytics-root.js]: Is the tick necessary here? Or can you just immediately resolve?

File: extensions/amp-analytics/0.1/analytics-root.js [hunk 2]
```diff
@@ -173,39 +173,45 @@ export class AnalyticsRoot {
...
    * analytics root in relationship to the specified context node.
-   * Note: Please wait for document-ready to avoid false missed searches with
-   * selector other than :root or :host
    *
    * @param {!Element} context
    * @param {string} selector DOM query selector.
    * @param {?string=} selectionMethod Allowed values are `null`,
    *   `'closest'` and `'scope'`.
-   * @return {?Element} Element corresponding to the selector if found.
+   * @return {Promise<?Element>} Element corresponding to the selector if found.
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/analytics-root.js]: It'll be `{!Promise<?Element>}`

File: extensions/amp-analytics/0.1/analytics-root.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/analytics-root.js]: No need. Just `return Promise.resolve(...)`

File: extensions/amp-analytics/0.1/analytics-root.js [hunk 3]
```diff
@@ -173,39 +173,45 @@ export class AnalyticsRoot {
...
+    // And no need to wait for document ready.
     if (selector == ':root') {
-      return this.getRootElement();
+      return Promise.resolve().then(() => {
+        return this.getRootElement();
+      });
     }
     if (selector == ':host') {
-      return this.getHostElement();
+      return Promise.resolve().then(() => {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/analytics-root.js]: Ditto, `return Promise.resolve(this.getHostElement())`

File: extensions/amp-analytics/0.1/analytics-root.js [hunk 4]
```diff
@@ -216,16 +222,17 @@ export class AnalyticsRoot {
    * @param {string} selector DOM query selector.
    * @param {?string=} selectionMethod Allowed values are `null`,
    *   `'closest'` and `'scope'`.
-   * @return {?AmpElement} AMP element corresponding to the selector if found.
+   * @return {Promise<?AmpElement>} AMP element corresponding to the selector if found.
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/analytics-root.js]: `!Promise`

File: extensions/amp-analytics/0.1/events.js [hunk 5]
```diff
@@ -162,8 +154,10 @@ export class CustomEventTracker extends EventTracker {
     }
 
     return this.observers_[eventType].add(event => {
+      //console.log('add listener');
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/events.js]: Remove

File: extensions/amp-analytics/0.1/events.js [hunk 6]
```diff
@@ -162,8 +154,10 @@ export class CustomEventTracker extends EventTracker {
...
       // Wait for target selected
       targetReady.then(target => {
+        //console.log('trigger target handler');
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/events.js]: Remove

File: extensions/amp-analytics/0.1/events.js [hunk 7]
```diff
@@ -339,17 +333,17 @@ export class IniLoadTracker extends EventTracker {
...
+      const selectionMethod = config['selectionMethod'];
+      promise = this.root.getAmpElement(
+          (context.parentElement || context),
+          selector,
+          selectionMethod
+          ).then(element => {
+            target = user().assertElement(
+                element, `Element "${selector}" not found`);;
+            return this.getElementSignal(
+                'ini-load', /** @type{!Element} */ (element));
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/events.js]: Pass `target` instead.

File: extensions/amp-analytics/0.1/events.js [hunk 8]
```diff
@@ -421,22 +415,22 @@ export class VisibilityTracker extends EventTracker {
...
-          this.getReadyPromise(waitForSpec, selector, element),
-          createReadyReportPromiseFunc,
-          this.onEvent_.bind(this, eventType, listener, element));
-    });
+    const unlistenPromise = this.root.getAmpElement(
+        (context.parentElement || context),
+        selector,
+        selectionMethod
+        ).then(element => {
+          element =
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/events.js]: Is it worth adding `getRequiredAmpElement` and `getRequiredElement()` APIs in the root? They seem to come up a lot. Or we can even modify the existing APIs if this is ALWAYS the case?

Comment (zhouyx) [inline extensions/amp-analytics/0.1/events.js]: I checked we always check if element is not null. Hence I moved to existing APIs. PTAL

File: extensions/amp-analytics/0.1/events.js (hunk 5 above)
Comment (zhouyx) [inline extensions/amp-analytics/0.1/events.js]: oops.

File: extensions/amp-analytics/0.1/analytics-root.js (hunk 3 above)
Comment (zhouyx) [inline extensions/amp-analytics/0.1/analytics-root.js]: I used `promise.catch(error)` in testing.
However when I use `return Promise.resolve(user.assertElement(this.getHostElement()))`, I was not able to catch the error. Then I changed to `Promise.resolve().then(() => return user.assertElement(this.getHostElement()))` and then was able to catch the error. Can someone explain why this is the case?

File: extensions/amp-analytics/0.1/analytics-root.js (hunk 2 above)
Comment (zhouyx) [inline extensions/amp-analytics/0.1/analytics-root.js]: done

File: extensions/amp-analytics/0.1/analytics-root.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/analytics-root.js]: I see. Slightly weird. But I agree - it's probably better to yield error consistently.

Comment (jridgewell) [inline extensions/amp-analytics/0.1/analytics-root.js]: If the `assertElement` is throwing, then put it in a Promise constructor:

```js
return new Promise(res => {
  resolve(user().assertElement(this.getRootElement()));
});
```

Comment (zhouyx) [inline extensions/amp-analytics/0.1/analytics-root.js]: changed as suggested : )

**Pull request merged**